### PR TITLE
Handle `find`, `count` and `delete` correctly when collection doesn't exist

### LINKED ISSUE #651: Calling a non-existing collection causes internal error
When querying a non-existing collection, we should do the same as MongoDB: return an empty result and do not create the collection/table. I think we used to do that, but now we don't.

We should also do the same thing as MongoDB for updates, deletes, etc.

* [x] Add tests for
  * [x] `count` in `query_test.go`
  * [x] `delete` in a new `delete_test.go`
  * [x] `find` in `query_test.go`
  * [x] `findAndModify` in `findandmodify_test.go`
  * [x] `update` in `update_test.go`
* [x] Fix code.

That should be done for both pg and tigris handlers. Use compatibility tests where possible.

The original report follows.

---

### How to reproduce

Version: [18daf68 on Fri May 27 20:51:37 2022 +0300](https://github.com/FerretDB/FerretDB/tree/18daf6891aeccbd507605402c01484299aaa2e2a)

1. Connect to the `test` database provided by `task env-up`.
2. Run the query `db.pampam.find();` (`pampam` collection doesn't exist in the `test` db)

### Actual result

The server returns an internal error and closes the connection.
Here is what I see in the logs:

```
2022-05-28T00:31:35.243+0200	DEBUG	// 127.0.0.1:59314 -> 127.0.0.1:27017 .// 127.0.0.1:59314 -> 127.0.0.1:27017 	clientconn/conn.go:166	Request header: length:    70, id: 3736, response_to:    0, opcode: OP_MSG
2022-05-28T00:31:35.243+0200	DEBUG	// 127.0.0.1:59314 -> 127.0.0.1:27017 .// 127.0.0.1:59314 -> 127.0.0.1:27017 	clientconn/conn.go:167	Request message:
{
  "Checksum": 0,
  "FlagBits": 0,
  "Sections": [
    {
      "Document": {
        "$k": [
          "find",
          "filter",
          "$db"
        ],
        "find": "pampam",
        "filter": {
          "$k": []
        },
        "$db": "test"
      },
      "Kind": 0
    }
  ]
}



2022-05-28T00:31:35.248+0200	ERROR	pg.Pool	v4@v4.16.1/conn.go:354	Query	{"err": "ERROR: relation \"test.pampam\" does not exist (SQLSTATE 42P01)", "sql": "SELECT _jsonb FROM \"test\".\"pampam\"", "args": [], "pid": 140}
github.com/jackc/pgx/v4.(*Conn).log
	github.com/jackc/pgx/v4@v4.16.1/conn.go:354
github.com/jackc/pgx/v4.(*connRows).Close
	github.com/jackc/pgx/v4@v4.16.1/rows.go:153
github.com/jackc/pgx/v4.(*connRows).fatal
	github.com/jackc/pgx/v4@v4.16.1/rows.go:178
github.com/jackc/pgx/v4.(*Conn).Query
	github.com/jackc/pgx/v4@v4.16.1/conn.go:601
github.com/jackc/pgx/v4/pgxpool.(*Conn).Query
	github.com/jackc/pgx/v4@v4.16.1/pgxpool/conn.go:54
github.com/jackc/pgx/v4/pgxpool.(*Pool).Query
	github.com/jackc/pgx/v4@v4.16.1/pgxpool/pool.go:496
github.com/FerretDB/FerretDB/internal/handlers/pg.(*Handler).fetch
	github.com/FerretDB/FerretDB/internal/handlers/pg/fetch.go:49
github.com/FerretDB/FerretDB/internal/handlers/pg.(*Handler).MsgFind
	github.com/FerretDB/FerretDB/internal/handlers/pg/msg_find.go:104
github.com/FerretDB/FerretDB/internal/clientconn.(*conn).handleOpMsg
	github.com/FerretDB/FerretDB/internal/clientconn/conn.go:386
github.com/FerretDB/FerretDB/internal/clientconn.(*conn).route
	github.com/FerretDB/FerretDB/internal/clientconn/conn.go:281
github.com/FerretDB/FerretDB/internal/clientconn.(*conn).run
	github.com/FerretDB/FerretDB/internal/clientconn/conn.go:175
github.com/FerretDB/FerretDB/internal/clientconn.(*Listener).Run.func2
	github.com/FerretDB/FerretDB/internal/clientconn/listener.go:139
2022-05-28T00:31:35.249+0200	DEBUG	// 127.0.0.1:59314 -> 127.0.0.1:27017 .// 127.0.0.1:59314 -> 127.0.0.1:27017 	clientconn/conn.go:373	Response header: length:   184, id:    2, response_to: 3736, opcode: OP_MSG
2022-05-28T00:31:35.250+0200	DEBUG	// 127.0.0.1:59314 -> 127.0.0.1:27017 .// 127.0.0.1:59314 -> 127.0.0.1:27017 	clientconn/conn.go:374	Response message:
{
  "Checksum": 0,
  "FlagBits": 0,
  "Sections": [
    {
      "Document": {
        "$k": [
          "ok",
          "errmsg",
          "code",
          "codeName"
        ],
        "ok": {
          "$f": 0
        },
        "errmsg": "[fetch.go:51 pg.(*Handler).fetch] ERROR: relation \"test.pampam\" does not exist (SQLSTATE 42P01)",
        "code": 1,
        "codeName": "InternalError"
      },
      "Kind": 0
    }
  ]
}



2022-05-28T00:31:35.256+0200	DEBUG	// 127.0.0.1:59314 -> 127.0.0.1:27017 .// 127.0.0.1:59314 -> 127.0.0.1:27017 	clientconn/conn.go:190	Proxy header: length:   100, id:  489, response_to: 3736, opcode: OP_MSG
2022-05-28T00:31:35.257+0200	DEBUG	// 127.0.0.1:59314 -> 127.0.0.1:27017 .// 127.0.0.1:59314 -> 127.0.0.1:27017 	clientconn/conn.go:191	Proxy message:
{
  "Checksum": 0,
  "FlagBits": 0,
  "Sections": [
    {
      "Document": {
        "$k": [
          "cursor",
          "ok"
        ],
        "cursor": {
          "$k": [
            "id",
            "ns",
            "firstBatch"
          ],
          "id": {
            "$l": "0"
          },
          "ns": "test.pampam",
          "firstBatch": []
        },
        "ok": {
          "$f": 1
        }
      },
      "Kind": 0
    }
  ]
}



2022-05-28T00:31:35.258+0200	INFO	// 127.0.0.1:59314 -> 127.0.0.1:27017 .// 127.0.0.1:59314 -> 127.0.0.1:27017 	clientconn/conn.go:221	Header diff:
--- res header
+++ proxy header
@@ -1 +1 @@
-length:   184, id:    2, response_to: 3736, opcode: OP_MSG
+length:   100, id:  489, response_to: 3736, opcode: OP_MSG

Body diff:
--- res body
+++ proxy body
@@ -7,13 +7,20 @@
         "$k": [
-          "ok",
-          "errmsg",
-          "code",
-          "codeName"
+          "cursor",
+          "ok"
         ],
+        "cursor": {
+          "$k": [
+            "id",
+            "ns",
+            "firstBatch"
+          ],
+          "id": {
+            "$l": "0"
+          },
+          "ns": "test.pampam",
+          "firstBatch": []
+        },
         "ok": {
-          "$f": 0
-        },
-        "errmsg": "[fetch.go:51 pg.(*Handler).fetch] ERROR: relation \"test.pampam\" does not exist (SQLSTATE 42P01)",
-        "code": 1,
-        "codeName": "InternalError"
+          "$f": 1
+        }
       },



2022-05-28T00:31:35.258+0200	WARN	listener	clientconn/listener.go:143	Connection stopped	{"error": "fatal error"}
github.com/FerretDB/FerretDB/internal/clientconn.(*Listener).Run.func2
	github.com/FerretDB/FerretDB/internal/clientconn/listener.go:143
```

### Expected result

If a client tries to find data in a collection that doesn't exist, it's not an internal error.

### PR DESCRIPTION
**Pull request by rumyantseva**
## Description

This PR doesn't solve #651 but covers a part of it.

Issue #651 is quite a big one, so I suggest dividing it into two PRs.

**In the scope of this PR:**
- `find` should return an empty response, with no errors.
- `count` should return 0.
- `delete` shouldn't return any errors.

**In a separate PR (to be done):**
- `findAndModify` (this method is tricky as it needs to create a doc that represents data from both `query` and `update` docs).
- `update` (same reason).

**Other fixes in the current PR:**
I made an ugly bug in #650, and in fact, proxy responses haven't been logged since then. I've just fixed that bug.

## Testing

- I added tests for `count`, `find`, and `delete` to cover cases when collection doesn't exist.
- In addition, I manually checked various cases to compare responses between "pg-based implementation" and "proxy" to make sure that the logic is the same.

## Checklist

- [x] Code is regenerated if needed (`task gen`).
- [x] Code is formatted (`task fmt`).
- [x] Docs are formatted (`task docs-fmt`).
- [x] Tests passed (`task test`).
- [x] Linters passed (`task lint`).
- [x] PR assignee is set.
- [x] PR labels are set.
- [x] Project and iternation are set.
- [ ] Critical codecov complaints are considered.
- [x] PR title and body are in accordance with the Conform PR GitHub Action.
- [x] Other GitHub Actions passed.

### CODE REVIEW

File: internal/clientconn/conn.go [hunk 1]
```diff
@@ -192,7 +192,7 @@ func (c *conn) run(ctx context.Context) (err error) {
 			}
 
 			proxyHeader, proxyBody, _ = c.proxy.Route(ctx, reqHeader, reqBody)
-			if level := c.logResponse(resHeader, resBody, resCloseConn); level != diffLogLevel {
+			if level := c.logResponse("Proxy response", proxyHeader, proxyBody, resCloseConn); level != diffLogLevel {
```
Comment (rumyantseva) [inline internal/clientconn/conn.go:195]: What a bug I produced in the logger task!

File: internal/handlers/pg/pgdb/pool.go [hunk 2]
```diff
@@ -378,6 +376,17 @@ func (pgPool *Pool) EnsureTableExist(ctx context.Context, db, collection string)
 	return true, nil
 }
 
+// TableExists returns true if given FerretDB database / PostgreSQL schema
+// and FerretDB collection / PostgreSQL table exist.
+func (pgPool *Pool) TableExists(ctx context.Context, db, collection string) (bool, error) {
```
Comment (rumyantseva) [inline internal/handlers/pg/pgdb/pool.go:381]: I'd love to write some (integration) tests for such functions, but I see there are no tests in `handlers/pg` nor in `pgdb`...

File: internal/handlers/pg/msg_update.go [hunk 3]
```diff
@@ -63,7 +63,7 @@ func (h *Handler) MsgUpdate(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg,
 		return nil, err
 	}
 
-	created, err := h.pgPool.EnsureTableExist(ctx, sp.db, sp.collection)
+	created, err := h.pgPool.CreateTableIfNotExists(ctx, sp.db, sp.collection)
```
Comment (rumyantseva) [inline internal/handlers/pg/msg_update.go:66]: Just a func rename. This method needs some special handling to deal with `request` correctly, but it's not in scope of this PR.

File: internal/handlers/pg/msg_insert.go [hunk 4]
```diff
@@ -98,6 +90,14 @@ func (h *Handler) MsgInsert(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg,
 
 // insert prepares and executes actual INSERT request to Postgres.
 func (h *Handler) insert(ctx context.Context, sp sqlParam, doc any) error {
+	created, err := h.pgPool.CreateTableIfNotExists(ctx, sp.db, sp.collection)
```
Comment (rumyantseva) [inline internal/handlers/pg/msg_insert.go:93]: Moved this code here for consistency, as `fetch` has a similar construction. Also, `upsert` will have it too.

File: internal/handlers/pg/msg_insert.go [hunk 5]
```diff
@@ -59,14 +59,6 @@ func (h *Handler) MsgInsert(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg,
 		return nil, err
 	}
 
-	created, err := h.pgPool.EnsureTableExist(ctx, sp.db, sp.collection)
```
Comment (rumyantseva) [inline internal/handlers/pg/msg_insert.go:62]: That was the only place in the code where a non-existing collection was handled.

**Review by seeforschauer — APPROVED** [review]

**Review by w84thesun — APPROVED** [review]

File: internal/handlers/pg/pgdb/pool.go [hunk 6]
```diff
@@ -348,16 +348,14 @@ func (pgPool *Pool) DropTable(ctx context.Context, schema, table string) error {
...
-	tables, err := pgPool.Tables(ctx, db)
-	if err != nil {
-		return false, lazyerrors.Error(err)
-	}
-
-	if slices.Contains(tables, collection) {
+func (pgPool *Pool) CreateTableIfNotExists(ctx context.Context, db, collection string) (bool, error) {
+	if exists, err := pgPool.TableExists(ctx, db, collection); err != nil {
+		return false, err
+	} else if exists {
```
Comment (AlekSi) [inline internal/handlers/pg/pgdb/pool.go:358]: ^

File: internal/handlers/pg/pgdb/pool.go [hunk 7]
```diff
@@ -348,16 +348,14 @@ func (pgPool *Pool) DropTable(ctx context.Context, schema, table string) error {
...
 		return false, nil
 	}
```
Comment (AlekSi) [inline internal/handlers/pg/pgdb/pool.go:360]:
```suggestion
	exists, err := pgPool.TableExists(ctx, db, collection)
	if err != nil {
		return false, lazyerrors.Error(err)
	}
	if exists {
		return false, nil
	}
```

File: internal/handlers/pg/pgdb/pool.go (hunk 2 above)
Comment (AlekSi) [reply internal/handlers/pg/pgdb/pool.go:381]: Let's do that

**Review by AlekSi — COMMENTED** [review]
And let's add tests for `TableExists` and `CreateTableIfNotExists`

**Pull request closed without merging**